### ISSUE TITLE
Optional outputs can cause legit partially-satisfied prereqiuisites

### ISSUE THREAD

**Issue by hjoliver** [cylc/cylc-flow#5729]
Simplified version of user workflow:
```ini
[scheduling]
   [[graph]]
      R1 = """
         FAM:fail-all? => bad
         FAM:succeed-any? => good
     """
[runtime]
   [[FAM]]
   [[a]]
      inherit = FAM
      script = true
   [[b]]
      inherit = FAM
      script = false
   [[bad]]
   [[good]]
```

Result:
```console
INFO - [1/good running job:01 flows:1] => succeeded
WARNING - Partially satisfied prerequisites:
      * 1/bad is waiting on ['1/a:failed']
CRITICAL - Workflow stalled
```

The problem is: *`bad` depends on multiple optional outputs, so we should not be surprised that it ends up partially satisfied.*

Currently the only automatic solution is ... our old friend the suicide trigger 😠 
```ini
"FAM:succeed-any? & FAM:finish-all? => !bad"
```
(We have to wait for all members of FAM to finish, because `bad` could be spawned after `good` triggers).

**Comment by hjoliver** [comment]
To refine this a bit, it's not specifically to do with family triggers.
```ini
"a? & b? => c  # with a fails and b succeeds"
```
or even:
```ini
"a? & b? => c  # with a fails and b succeeds"
```
This will stall with `c` partially satisfied. 

The graph says, trigger `c` only if both `a` and `b` succeed, *and* they are expected to fail sometimes - so that's not a reason to assume an error has occurred.

**Comment by hjoliver** [comment]
Assigned to 8.x milestone, but we should bring it forward it if possible - it's a bit nasty for users to understand and deal with.

**Comment by oliver-sanders** [comment]
Sticking the question label on this to flag for our next meeting as this will probably require some hashing out of details.

**Comment by oliver-sanders** [comment]
Recently discovered example:

```
a? & b => c
```

It's not clear exactly what the user wanted in the `a:fail` case, however, a stall apparently wasn't what they were going for.

**Comment by oliver-sanders** [comment]
Documented this limitation and the suicide trigger workaround in https://github.com/cylc/cylc-doc/pull/772

**Comment by hjoliver** [comment]
Documenting this is good for now, but I still think, generally, we should not stall if the reason for the stall is ONLY unsatisfied optional outputs. It shouldn't need suicide triggers. I think from memory of the initial discussion about this @dpmatthews did not like this suggestion? If so, what's the reason?

**Comment by oliver-sanders** [comment]
I don't think there is any resistance to solving this issue if it is possible to do so, but there is concern that it might not be possible to resolve this issue safely without undermining other use cases.

This issue isn't going to be resolved for a while so we need to at least document it for now and see what can be done for it when the time allows.

**Comment by dpmatthews** [comment]
In my view, `a? => c` implies "if a succeeds then c must be run (when all it's pre-requisites are met)".
If `b? => c` gets added to the graph (possibly by someone else in a completely different part of the graph) that shouldn't mean c might not run - that wouldn't be safe.

This doesn't just apply to optional outputs - the same can happen with optional branches.

If this is seen as a  major problem then I think we'd need to introduce a new syntax, e.g.
`a? =? c` implies "if a succeeds then run c if all it's pre-requisites are met"

**Comment by oliver-sanders** [comment]
> If this is seen as a major problem then I think we'd need to introduce a new syntax

Modelling "optionality" on the edge (i.e. trigger) was my initial suggestion (https://github.com/cylc/cylc-admin/pull/128) which was rejected in favour of modelling on the node (i.e. output) (https://github.com/cylc/cylc-admin/pull/129).

I think there's a lot to be said for modelling on the edge (e.g. access to the full power of conditional statements) and that it's probably a more "natural" paradigm for graph designers (users frequently get mixed up with optional outputs in a way that suggests they are thinking in triggers not outputs, e.g. why do I need a question mark on the RHS? Or, do I need a question mark on an optional task?).

However, migrating at this point is simply not feasible and supporting both concepts in parallel would make a confusing situation much worse (not to mention the challenging interaction space between the two). Easier just to document the suicide trigger approach.

**Comment by hjoliver** [comment]
Optional triggers was nice in some ways, but more confusing in others, e.g: https://github.com/cylc/cylc-admin/pull/128#issuecomment-886303145

> In my view, a? => c implies "if a succeeds then c must be run (when all it's pre-requisites are met)".
If b? => c gets added to the graph (possibly by someone else in a completely different part of the graph) that shouldn't mean c might not run - that wouldn't be safe.

I can't say I agree with this @dpmatthews.

We always have to take *all* of a task's triggers into account, and they're often spread all over the graph. We should not base behaviour on an assumption that the actual graph might not represent the user's true intentions.

By definition, if an output is optional *it doesn't have to be generated*, so logically a task that depends on it *doesn't have to run*. That also goes for a task that depends on multiple optional outputs, If a stall results solely from optional outputs we should remove the task automatically and log the reason.

An unnecessary stall, when the graph does in fact reflect the user's intentions, is also a bad outcome (it can, for instance, completely halt the progress of an experiment over the weekend, or worse).

**Comment by dpmatthews** [comment]
>I can't say I agree with this @dpmatthews. a? => c means "c depends on the (optional) success of a" end of story. Putting your spin on top of that is an optional extra that I don't recall being agreed.

My description simply describes the way Cylc has always worked - partially satisfied pre-requisites have always results in a stall.

> By definition, if an output is optional it doesn't have to be generated, so logically a task that depends on it doesn't have to run.

If an output is required it doesn't have to be generated either (the task might not run). I don't think the fact that an output is marked as optional makes any difference.

**Comment by hjoliver** [comment]
@dpmatthews - as per the wording of my original examples above (and what I just said in my previous post) I am assuming that the parent tasks - i.e. the owners of the optional outputs - actually ran, and I think you're worried about cases were the parents tasks didn't run at all due to events further upstream.

Let's make the distinction explicit then, and it should satisfy both of us:

If the workflow stalls with a partially satisfied task present, examine its unsatisfied prerequisites:
- IF they are all optional outputs AND all the parent tasks ran to completion, remove the task (and unstall the workflow)
- otherwise (not optional outputs, or the parent tasks did not run at all) do not remove the task (and remain stalled)

Do you agree?

**Comment by hjoliver** [comment]
> (note that submit-failed and expired are also "final" states, so the first example is not safe under the suggested logic above)

By "ran to completion" I mean fulfilled its completion condition, not just achieved a final status - so it is safe.

**Comment by hjoliver** [comment]
Let's get back to the basics.

Presumably your objections must boil down to this:
- *it is not safe to remove a partially satisfied task because we can't be sure that those prerequisites won't get satisfied in the future*

Is that right?

Well, I think we should be able to definitively identify cases where we can be sure. For example, simplest case:

```
a:x? & a:y? => b 

# Result:
# 1. a ran and succeeded, generating output :x but not :y
#   (so b is partially satisfied, stuck in the pool waiting on a:y?)
# 2. b has no other triggers
```

If 1.-2. are true, then actually we wouldn't even have to wait for a stall to remove `b`. 
- the graph says *"run `b` ONLY if `a:x?` AND `a:y?` are both generated"*
  - (and being optional, they might not both be generated - and that's perfectly OK)
- `a` ran successfully and fulfilled its completion condition in the workflow (i.e., nothing went wrong)
- but `b`'s trigger condition was not satisfied
- therefore `b` should not be triggered

In this particular case at least, everything went as planned and there is clearly no reason to stall the workflow. 
Not to remove `b` automatically is to cause an entirely unnecessary stall, which is a bad outcome for users.

[Aside: all else being equal, why should this stall when the single trigger case doesn't? `a:y? => b`]
 
@dpmatthews and/or @oliver-sanders - can you please respond specifically to this example:
- if you do not agree, what specifically do you think is wrong with my logic in respect to this example?
- (if you do agree with my logic for this simple case, we can move onto more general cases)

**Comment by hjoliver** [comment]
(Note we can avoid considering unsatisfied prerequisites where the parent tasks have not run to completion - if the workflow has not stalled yet, they may yet get satisfied; if the workflow has stalled due to presence of final incomplete tasks - that is an error condition and fixing it - by completing those tasks - may result in satisfying the prerequisites once the flow gets going again).

**Comment by oliver-sanders** [comment]
Sorry, my example was bunk. If I can wind back a couple of posts...


Here's a nasty example that highlights the problem:

```cylc
[scheduling]
    [[graph]]
        R1 = """
            a:x? & b:x? => x
            a:y? & b:y? => y

            x | y => c
        """

[runtime]
    [[a, b]]
        completion = succeeded and (x or y)
        [[[outputs]]]
            x = xxx
            y = yyy
    [[x, y]]
    [[c]]
```

We have two alternate pathways through the graph. The user wants Cylc to follow one of them.

User's intention:

| | a:x | a:y |
| --- | --- | --- |
| b:x | run "x" | stall |
| b:y | stall | run "y" |

Cylc's current behaviour matches the user's intention.


---

> ```cylc
> a:x? & a:y? => b 
>
> # Result:
> # 1. a ran and succeeded, generating output :x but not :y
> #   (so b is partially satisfied, stuck in the pool waiting on a:y?)
> # 2. b has no other triggers
> ```

Note, this example matches your `a:x? & a:y? => b` example above, however, as I was explaining in my initial post, the user's intention is completely different. In this case you want:

| | a:x | a:y |
| --- | --- | --- |
| b:x | run "x" | do-nothing |
| b:y | do-nothing | run "y" |

Under current logic, this intention can be expressed like so (the newly documented workaround):

```cylc
[scheduling]
    [[graph]]
        R1 = """
            a:x? & b:x? => x
            a:y? & b:y? => y

            a:x? | b:x? => !y
            a:y? | b:y? => !x

            x | y => c
        """
```

---

So, there are two possible intentions behind `a:x? & a:y? => b`.

You have assumed that having one, but not both outputs complete means the downstream task is not needed in **any** case (what I meant by "optional task"), however, this scenario can also indicate a genuine graph evolution error (what I meant by "required task").

We cannot tell the user's intention without more information then is provided by optional outputs alone. As a result, I don't think that this problem can be solved under this model, hence Dave's suggestion of using optional triggers to provide this required context.

---

> Presumably your objection must boil down to one of these?:
> * it is never safe to remove a partially satisfied task because we can't be sure that it won't get satisfied in the future, or
> * it is never safe to remove a partially satisfied task because it always indicates an error, or at least can't easily be >       distinguished from error cases, that must be highlighted

Again, there's no "objection" to solving this, I'm just not convinced that it is solvable.

I don't agree in full with either of the bullets above, (it is sometimes safe to remove a partially satisfied task, but, other times it is not). I would say:

* It is not safe to remove a partially satisfied task because we cannot tell if it indicates an error or not.

**Comment by hjoliver** [comment]
OK thanks for that, it makes a lot more sense now - the "bunk" example was really screwing with my head!

On reflection, it seems our respective perspectives boil down to this:

------
*There are possible combinations of outputs that the graph does not explicitly handle. If such a combination eventuates at runtime and leaves a task with partially satisfied prerequisites:*
- *NZ: the graph does not say to do anything in this situation, so don't do anything*
   - *(just back out of the pre-spawning, for prerequisite tracking, which created a task that turned out not to be needed)*
- *UK: the graph does not say what to do in this situation, so stall the workflow*
   - *(the user might have actually intended a stall in this situation)*
-----

*Maybe* these two approaches are are just a matter of preference, in which case unfortunately for me it seems I'm gonna be outvoted, but I still don't think your approach is the right one. Various reasons follow:

This does not strike me as "graph evolution error" at all. I'm not assuming anything about the user's intention, just taking their graph at its word: if it says to trigger certain tasks in response to certain events, do that, end of story. 

> We have two alternate pathways through the graph. The user wants Cylc to follow one of them.

The user has given two pathways for two specific (combinations of) events, but according to their own workflow configuration there are other possible events as well - how can they reasonably want Cylc to follow one of the two paths even if the specific events that trigger those paths do not occur?!  If you have binary alternate graph paths and you want one or the other to run every time, they had better be predicated by binary events.

> User's intention: ... stall

Since when does a user "intend" their workflow to stall? A stall should indicate that something has gone wrong that requires user intervention to fix. In your table of outputs and intentions, "stall" should be replaced by what the user actually wants to happen if that combination of events occurs? Can you give a concrete example where a stall is the true intention, and what the intervention would be to recover from that stall?

I think you're really making the claim that every possible combination of events must be handled explicitly, and if anything else happens we should assume that the user didn't anticipate it and might want a different graph if they had (vs they simply did not add graph paths for events that they don't care about). 

But that is inconsistent because you're using partially satisfied prerequisites as a proxy for a situation that can also easily occur without them: 
```
a:x? => x
# a generates :y at runtime but not :x
# there are no graph paths defined for :y
# this does NOT cause a stall
```

**Comment by dwsutherland** [comment]
I'm probably with Hilary, and as he hinted, this wouldn't be an issue if we didn't spawn partially satisfied tasks.

If we only spawned parentless and prereq satisfied tasks the examples would run as they read, and the stall intentions would seem a little silly in this context.. 
But perhaps users expect something different with the practical concession of spawning partially satisfied tasks.

**Comment by hjoliver** [comment]
https://cylc.discourse.group/t/graph-branching-vs-suicide-trigger/1111

**Comment by dwsutherland** [comment]
> If we only spawned parentless and prereq satisfied tasks

Obviously, this would require a different option for knowing when a task can run.. Something like, given a task output lookup/generate what tasks have it as a prerequisite and check all other prerequisites (from a DB of sorts) of these tasks, if satisfied run.

The partially satisfied tasks would turn up in the `n=1` data-store window, and wouldn't be in the active pool to need removal if a branch isn't taken.

**Comment by hjoliver** [comment]
> Obviously, this would require a different strategy for knowing when a task can run..

I'm in danger of confusing the issue by going off topic here, but I don't think you mean "knowing when a task can run". 

I think of spawning on the first output as merely an easy way (given Cylc internals) of tracking - and exposing to users - partially satisfied prerequisites. It doesn't make any difference to when tasks can run.

> (if xtriggers were represented as running tasks, then we wouldn't have any waiting tasks at n=0)

Even better, an xtrigger (internal to the scheduler, no need to represent as a running task) should spawn dependent tasks on demand (i.e., at the moment the external condition becomes satisfied), rather than belonging to pre-spawned waiting tasks.  Same result though, in terms of less waiting tasks in `n=0`.

**Comment by dwsutherland** [comment]
> I'm in danger of confusing the issue by going off topic here

I was merely reducing our spawning strategy down to it's idealized limit to illustrate that stalling doesn't make sense for these optional output limits.

> Even better, an xtrigger (internal to the scheduler, no need to represent as a running task) should spawn dependent tasks on demand

Yeah, only problem is you can't expand a window (to inform the user) about no tasks.

**Comment by hjoliver** [comment]
> I was merely reducing our spawning strategy down to it's idealized limit to illustrate that stalling doesn't make sense for these optional output examples.

I agree! - but that doesn't change "knowing when a task can *run*" (maybe you didn't mean to use "run" there?)

> Yeah, only problem is you can't expand a window (to inform the user) about no tasks.

Yes, we'd need a different strategy for exposing the right info to users.  The current way, despite being a sort of kludge that leveraged existing Cylc internals, does have some advantages.

**Comment by oliver-sanders** [comment]
> Relatedly, if the downstream task gets zero prerequisites satisfied rather than some-but-not-all, the workflow may shut down rather than stall:

Yes, that example is the mirror opposite of this issue.

The solution:

```cylc
R1 = (a:finished & b:finished) and (a? | b?) => c
```

Or as it's perhaps more likely to be written:

```cylc
R1 = POLLER:finish-all & POLLER:succeed-any => c
```

Logical from an SoD perspective, but not the best from the user's perspective.

There's a limit to what we can achieve under the optional outputs model, there's just too much user intention to denote with a single `?`.

Stalling is generally safer/better than shutting down as it tells the user that something is wrong and gives them the chance to fix it, so we tried to engineer stalling as the default where possible.

This is a situation is one where we couldn't make stalling the default, and accepted the need for an "artificial dependency" to prevent premature shutdown:

> `a => c # ARTIFICIAL DEPENDENCY`
>
> https://cylc.github.io/cylc-admin/proposal-new-output-syntax.html#path-branching

**Comment by hjoliver** [comment]
Yes you are right - I had forgotten that I myself had explicitly acknowledged the need for an artificial dependency to handle the "at least one task must succeed" use case, in the optional outputs proposal: *"Note that an artificial dependency can be used to ensure that at least one of the alternate paths gets taken:"*.

My perspective on this now is a bit different though. Saying that a task is "required" is really a constraint on the collective behaviour of its parents - i.e., the outputs it depends on must always conspire to trigger the task (which isn't necessarily the case for optional outputs). And if something goes wrong with that, it is the parents that should be presented to users as the problem.

> Logical from an SoD perspective, but not the best from the user's perspective.

It's not even logical from an "SoD perspective", unless you mean "our current implementation of SoD" (which has several deficiencies) as opposed to the principle of it. The polling use case makes it clear:

```
poller1? | poller2? => downstream
```
where the user knows one of the pollers must succeed. If both fail then (a) we didn't poll for long enough; or (b) the poller tasks have a bug; or (c) the upstream event will never happen.

All of the possibilities mean something is wrong with the poller tasks, not with `downstream`. (That's true even for (c) - it is wrong to poll for an event that will never happen, although manual intervention is required to inject that info). 

So, currently have to add an artificial dependency to make `c` spawn into `n=0` as incomplete to avoid an early shutdown, i.e. to use `c` as indicator of problems in the upstream poller tasks, which is ugly.

That's not a problem with with SoD concept, it's just shows our current optional outputs model can't handle the user expectation here, which is *"one of the poller tasks must succeed"* (or else something is wrong with the poller tasks).

Hence: https://github.com/cylc/cylc-flow/issues/7264
